### ISSUE TITLE
does promtail support Subdirectory

### ISSUE THREAD

**Issue by hellojukay** [grafana/loki#201]
does promtail support Subdirectory? if it does, how to config it.

**Comment by slim-bean** [comment]
PR #206 did add subdirectory support however there are a few issues.  Some of which will be resolved when I finish #293 but not all.

The glob support will allow matching something like `/var/log/*.log` and any file in `/var/log` that ends in `.log` would get tailed, however, no subdirectories would get picked up.

However if you do something like `/var/log/*/*.log` then any file ending in .log in any sub directories would get tailed, however, no files in `/var/log/` ending in .log would be tailed

Unfortunately there is no `**` matcher support with the filepath package, and they have a longstanding issue open which leads me to believe there never will be support : https://github.com/golang/go/issues/11862

This leaves us with a few options:

- leave things the way they are and to tail any file in a subdirectory you would need to make sure you have a wildcard somewhere in the path, e.g. `/var/log/*/...`
- write code that automatically searches subdirectories and include a flag to enable/disable this behavior, i.e. `/var/log/*.log` with the "search_subdirectores" flag set to true would recursively search for any .log files in any subdir in `/var/log`, with the flag false it would only tail files in `/var/log/` ending in .log
- investigate another library to do the matching which supports `**` style matches, there are a few mentioned in that issue above, though I'm not sure if any are really suitable.

I'm inclined to suggest the first option though in many cases it would require creating multiple configs to scrape `/var/log/*.log` and `/var/log/*/*.log` and I'm not sure if this would lead to any other issues.

@tomwilkie @woodsaj  thoughts?

**Comment by slim-bean** [comment]
After some discussion we have decided to move forward with a third party library.  I narrowed the choice down to two 

- https://github.com/mattn/go-zglob
- https://github.com/bmatcuk/doublestar

Choosing the second: https://github.com/bmatcuk/doublestar

It was not easy choosing between the two, both libraries are recently maintained, have test coverage, and have responsive authors.  My decision was based mostly on the doublestar library having what I believe to be a little better test coverage, a little better documentation, and some issues/responses that lead me to believe it's being used/works on Windows.
